File: ts/src/grammar.ts
/**
 * Social grammar module — high-level vertex operations that create properly
 * hash-chained events on the store. Ports the vertex operations from the Go
 * reference implementation.
 */
import { createEvent, type Signer, type Event } from "./event.js";
import type { Store } from "./store.js";
import { ActorId, ConversationId, EventId, EventType, Hash } from "./types.js";

/** Grammar wraps a Store and provides high-level social grammar operations. */
export class Grammar {
  constructor(private readonly store: Store) {}

  // ── helpers ──────────────────────────────────────────────────────────

  private headPrevHash(): Hash {
    const head = this.store.head();
    return head.isSome ? head.unwrap().hash : Hash.zero();
  }

  private record(
    eventType: EventType,
    source: ActorId,
    content: Record<string, unknown>,
    causes: EventId[],
    conversationId: ConversationId,
    signer: Signer,
  ): Event {
    const prevHash = this.headPrevHash();
    return this.store.append(
      createEvent(eventType, source, content, causes, conversationId, prevHash, signer),
    );
  }

  // ── vertex operations ────────────────────────────────────────────────

  /** Emit creates independent content. Causes must be non-empty. */
  emit(
    source: ActorId,
    body: string,
    conversationId: ConversationId,
    causes: EventId[],
    signer: Signer,
  ): Event {
    if (causes.length === 0) {
      throw new Error("emit requires at least one cause");
    }
    return this.record(
      new EventType("grammar.emitted"), source,
      { Body: body },
      causes, conversationId, signer,
    );
  }

  /** Respond creates causally dependent, subordinate content. */
  respond(
    source: ActorId,
    body: string,
    parent: EventId,
    conversationId: ConversationId,
    signer: Signer,
  ): Event {
    return this.record(
      new EventType("grammar.responded"), source,
      { Body: body, Parent: parent.value },
      [parent], conversationId, signer,
    );
  }

  /** Derive creates causally dependent but independent content. */
  derive(
    source: ActorId,
    body: string,
    sourceEvent: EventId,
    conversationId: ConversationId,
    signer: Signer,
  ): Event {
    return this.record(
      new EventType("grammar.derived"), source,
      { Body: body, Source: sourceEvent.value },
      [sourceEvent], conversationId, signer,
    );
  }

  /** Extend creates sequential content from the same author. */
  extend(
    source: ActorId,
    body: string,
    previous: EventId,
    conversationId: ConversationId,
    signer: Signer,
  ): Event {
    return this.record(
      new EventType("grammar.extended"), source,
      { Body: body, Previous: previous.value },
      [previous], conversationId, signer,
    );
  }

  /** Retract tombstones own content. Only the original author can retract. */
  retract(
    source: ActorId,
    target: EventId,
    reason: string,
    conversationId: ConversationId,
    signer: Signer,
  ): Event {
    const targetEvent = this.store.get(target);
    if (targetEvent.source.value !== source.value) {
      throw new Error(
        `retract: actor ${source.value} cannot retract event ${target.value} authored by ${targetEvent.source.value}`,
      );
    }
    return this.record(
      new EventType("grammar.retracted"), source,
      { Target: target.value, Reason: reason },
      [target], conversationId, signer,
    );
  }

  /** Annotate attaches metadata to existing content. */
  annotate(
    source: ActorId,
    target: EventId,
    key: string,
    value: string,
    conversationId: ConversationId,
    signer: Signer,
  ): Event {
    return this.record(
      new EventType("grammar.annotated"), source,
      { Target: target.value, Key: key, Value: value },
      [target], conversationId, signer,
    );
  }

  /** Merge joins two or more independent subtrees. Requires >= 2 sources. */
  merge(
    source: ActorId,
    body: string,
    sources: EventId[],
    conversationId: ConversationId,
    signer: Signer,
  ): Event {
    if (sources.length < 2) {
      throw new Error("merge requires at least two sources");
    }
    return this.record(
      new EventType("grammar.merged"), source,
      { Body: body, Sources: sources.map((s) => s.value) },
      sources, conversationId, signer,
    );
  }
}

File: rust/src/grammar.rs
use std::collections::BTreeMap;

use serde_json::Value;

use crate::errors::{EventGraphError, Result};
use crate::event::{create_event, Event, Signer};
use crate::store::{InMemoryStore, Store};
use crate::types::{ActorId, ConversationId, EventId, EventType, Hash};

/// High-level social grammar operations that create properly hash-chained events.
///
/// Each operation gets the current chain head, creates an event with the
/// appropriate type and content, appends it to the store, and returns it.
pub struct Grammar<'a> {
    store: &'a mut InMemoryStore,
}

impl<'a> Grammar<'a> {
    pub fn new(store: &'a mut InMemoryStore) -> Self {
        Self { store }
    }

    /// Returns the prev_hash for the next event: head's hash or zero if empty.
    fn prev_hash(&self) -> Hash {
        self.store
            .head()
            .map(|e| e.hash.clone())
            .unwrap_or_else(Hash::zero)
    }

    /// Emit a new utterance with explicit causal links.
    ///
    /// Type: `grammar.emitted`
    /// Content: `{"Body": body}`
    /// Causes must be non-empty.
    pub fn emit(
        &mut self,
        source: ActorId,
        body: &str,
        conversation_id: ConversationId,
        causes: Vec<EventId>,
        signer: &dyn Signer,
    ) -> Result<Event> {
        if causes.is_empty() {
            return Err(EventGraphError::GrammarViolation {
                detail: "emit requires at least one cause".to_string(),
            });
        }

        let mut content = BTreeMap::new();
        content.insert("Body".to_string(), Value::String(body.to_string()));

        let prev_hash = self.prev_hash();
        let event = create_event(
            EventType::new("grammar.emitted")?,
            source,
            content,
            causes,
            conversation_id,
            prev_hash,
            signer,
            1,
        );
        self.store.append(event)
    }

    /// Respond to a specific event.
    ///
    /// Type: `grammar.responded`
    /// Content: `{"Body": body, "Parent": parent_id}`
    /// Causes: `[parent]`
    pub fn respond(
        &mut self,
        source: ActorId,
        body: &str,
        parent: EventId,
        conversation_id: ConversationId,
        signer: &dyn Signer,
    ) -> Result<Event> {
        let mut content = BTreeMap::new();
        content.insert("Body".to_string(), Value::String(body.to_string()));
        content.insert(
            "Parent".to_string(),
            Value::String(parent.value().to_string()),
        );

        let prev_hash = self.prev_hash();
        let event = create_event(
            EventType::new("grammar.responded")?,
            source,
            content,
            vec![parent],
            conversation_id,
            prev_hash,
            signer,
            1,
        );
        self.store.append(event)
    }

    /// Derive a new statement from an existing event.
    ///
    /// Type: `grammar.derived`
    /// Content: `{"Body": body, "Source": source_event_id}`
    /// Causes: `[source_event]`
    pub fn derive(
        &mut self,
        source: ActorId,
        body: &str,
        source_event: EventId,
        conversation_id: ConversationId,
        signer: &dyn Signer,
    ) -> Result<Event> {
        let mut content = BTreeMap::new();
        content.insert("Body".to_string(), Value::String(body.to_string()));
        content.insert(
            "Source".to_string(),
            Value::String(source_event.value().to_string()),
        );

        let prev_hash = self.prev_hash();
        let event = create_event(
            EventType::new("grammar.derived")?,
            source,
            content,
            vec![source_event],
            conversation_id,
            prev_hash,
            signer,
            1,
        );
        self.store.append(event)
    }

    /// Extend a previous statement with additional content.
    ///
    /// Type: `grammar.extended`
    /// Content: `{"Body": body, "Previous": previous_id}`
    /// Causes: `[previous]`
    pub fn extend(
        &mut self,
        source: ActorId,
        body: &str,
        previous: EventId,
        conversation_id: ConversationId,
        signer: &dyn Signer,
    ) -> Result<Event> {
        let mut content = BTreeMap::new();
        content.insert("Body".to_string(), Value::String(body.to_string()));
        content.insert(
            "Previous".to_string(),
            Value::String(previous.value().to_string()),
        );

        let prev_hash = self.prev_hash();
        let event = create_event(
            EventType::new("grammar.extended")?,
            source,
            content,
            vec![previous],
            conversation_id,
            prev_hash,
            signer,
            1,
        );
        self.store.append(event)
    }

    /// Retract a previously authored event.
    ///
    /// Type: `grammar.retracted`
    /// Content: `{"Target": target_id, "Reason": reason}`
    /// Causes: `[target]`
    ///
    /// The source must be the author of the target event.
    pub fn retract(
        &mut self,
        source: ActorId,
        target: EventId,
        reason: &str,
        conversation_id: ConversationId,
        signer: &dyn Signer,
    ) -> Result<Event> {
        // Verify source authored the target event
        let target_event = self.store.get(&target)?;
        if target_event.source.value() != source.value() {
            return Err(EventGraphError::GrammarViolation {
                detail: format!(
                    "actor {} cannot retract event authored by {}",
                    source.value(),
                    target_event.source.value()
                ),
            });
        }

        let mut content = BTreeMap::new();
        content.insert(
            "Target".to_string(),
            Value::String(target.value().to_string()),
        );
        content.insert("Reason".to_string(), Value::String(reason.to_string()));

        let prev_hash = self.prev_hash();
        let event = create_event(
            EventType::new("grammar.retracted")?,
            source,
            content,
            vec![target],
            conversation_id,
            prev_hash,
            signer,
            1,
        );
        self.store.append(event)
    }

    /// Annotate an existing event with a key-value pair.
    ///
    /// Type: `grammar.annotated`
    /// Content: `{"Target": target_id, "Key": key, "Value": value}`
    /// Causes: `[target]`
    pub fn annotate(
        &mut self,
        source: ActorId,
        target: EventId,
        key: &str,
        value: &str,
        conversation_id: ConversationId,
        signer: &dyn Signer,
    ) -> Result<Event> {
        let mut content = BTreeMap::new();
        content.insert(
            "Target".to_string(),
            Value::String(target.value().to_string()),
        );
        content.insert("Key".to_string(), Value::String(key.to_string()));
        content.insert("Value".to_string(), Value::String(value.to_string()));

        let prev_hash = self.prev_hash();
        let event = create_event(
            EventType::new("grammar.annotated")?,
            source,
            content,
            vec![target],
            conversation_id,
            prev_hash,
            signer,
            1,
        );
        self.store.append(event)
    }

    /// Merge multiple events into a single synthesised statement.
    ///
    /// Type: `grammar.merged`
    /// Content: `{"Body": body, "Sources": [source_ids...]}`
    /// Causes: sources (must be >= 2)
    pub fn merge(
        &mut self,
        source: ActorId,
        body: &str,
        sources: Vec<EventId>,
        conversation_id: ConversationId,
        signer: &dyn Signer,
    ) -> Result<Event> {
        if sources.len() < 2 {
            return Err(EventGraphError::GrammarViolation {
                detail: "merge requires at least two source events".to_string(),
            });
        }

        let source_values: Vec<Value> = sources
            .iter()
            .map(|s| Value::String(s.value().to_string()))
            .collect();

        let mut content = BTreeMap::new();
        content.insert("Body".to_string(), Value::String(body.to_string()));
        content.insert("Sources".to_string(), Value::Array(source_values));

        let prev_hash = self.prev_hash();
        let event = create_event(
            EventType::new("grammar.merged")?,
            source,
            content,
            sources,
            conversation_id,
            prev_hash,
            signer,
            1,
        );
        self.store.append(event)
    }
}
